Add item requests

sashajaaa opened the pull request.

sashajaaa committed:
Commit №15.1:
feat:
-добавлена пагинация;
-добавлена реализация запросов и ответов (Item requests);
-добавлены тесты.

refactor:
-переработан файл pom.xml;
-переработан файл schema.sql;
-переработан класс BookingController;
-переработан класс BookingMapper;
-переработан интерфейс BookingRepository;
-добавлен интерфейс BookingService;
-переработан класс BookingServiceImpl;
-переработан класс CommentMapper;
-переработан класс ItemController;
-переработан класс ItemDto;
-переработан класс ItemMapper;
-переработан класс Item;
-переработан класс ItemRepository;
-добавлен интерфейс ItemService;
-переработан класс ItemServiceImpl;
-переработан класс ItemRequestController;
-переработан класс ItemRequestDto;
-добавлен класс ItemRequestMapper;
-переработан класс ItemRequest;
-добавлен интерфейс ItemRequestRepository;
-добавлен интерфейс ItemRequestService;
-добавлен класс ItemRequestServiceImpl;
-переработан класс UserController;
-переработан класс UserMapper;
-переработан класс UserService;
-добавлен интерфейс UserService;
-переработан класс UserServiceImpl.

sashajaaa committed:
Commit №15.1:
feat:
-добавлена пагинация;
-добавлена реализация запросов и ответов (Item requests);
-добавлены тесты.

refactor:
-переработан файл pom.xml;
-переработан файл schema.sql;
-переработан класс BookingController;
-переработан класс BookingMapper;
-переработан интерфейс BookingRepository;
-добавлен интерфейс BookingService;
-переработан класс BookingServiceImpl;
-переработан класс CommentMapper;
-переработан класс ItemController;
-переработан класс ItemDto;
-переработан класс ItemMapper;
-переработан класс Item;
-переработан класс ItemRepository;
-добавлен интерфейс ItemService;
-переработан класс ItemServiceImpl;
-переработан класс ItemRequestController;
-переработан класс ItemRequestDto;
-добавлен класс ItemRequestMapper;
-переработан класс ItemRequest;
-добавлен интерфейс ItemRequestRepository;
-добавлен интерфейс ItemRequestService;
-добавлен класс ItemRequestServiceImpl;
-переработан класс UserController;
-переработан класс UserMapper;
-переработан класс UserService;
-добавлен интерфейс UserService;
-переработан класс UserServiceImpl.

sashajaaa committed:
Commit №15.1:
feat:
-добавлена пагинация;
-добавлена реализация запросов и ответов (Item requests);
-добавлены тесты.

refactor:
-переработан файл pom.xml;
-переработан файл schema.sql;
-переработан класс BookingController;
-переработан класс BookingMapper;
-переработан интерфейс BookingRepository;
-добавлен интерфейс BookingService;
-переработан класс BookingServiceImpl;
-переработан класс CommentMapper;
-переработан класс ItemController;
-переработан класс ItemDto;
-переработан класс ItemMapper;
-переработан класс Item;
-переработан класс ItemRepository;
-добавлен интерфейс ItemService;
-переработан класс ItemServiceImpl;
-переработан класс ItemRequestController;
-переработан класс ItemRequestDto;
-добавлен класс ItemRequestMapper;
-переработан класс ItemRequest;
-добавлен интерфейс ItemRequestRepository;
-добавлен интерфейс ItemRequestService;
-добавлен класс ItemRequestServiceImpl;
-переработан класс UserController;
-переработан класс UserMapper;
-переработан класс UserService;
-добавлен интерфейс UserService;
-переработан класс UserServiceImpl.

sashajaaa committed:
Commit №15.1:
feat:
-добавлена пагинация;
-добавлена реализация запросов и ответов (Item requests);
-добавлены тесты.

refactor:
-переработан файл pom.xml;
-переработан файл schema.sql;
-переработан класс BookingController;
-переработан класс BookingMapper;
-переработан интерфейс BookingRepository;
-добавлен интерфейс BookingService;
-переработан класс BookingServiceImpl;
-переработан класс CommentMapper;
-переработан класс ItemController;
-переработан класс ItemDto;
-переработан класс ItemMapper;
-переработан класс Item;
-переработан класс ItemRepository;
-добавлен интерфейс ItemService;
-переработан класс ItemServiceImpl;
-переработан класс ItemRequestController;
-переработан класс ItemRequestDto;
-добавлен класс ItemRequestMapper;
-переработан класс ItemRequest;
-добавлен интерфейс ItemRequestRepository;
-добавлен интерфейс ItemRequestService;
-добавлен класс ItemRequestServiceImpl;
-переработан класс UserController;
-переработан класс UserMapper;
-переработан класс UserService;
-добавлен интерфейс UserService;
-переработан класс UserServiceImpl.

kalitva commented on pom.xml:
```diff
@@ -165,17 +186,17 @@
 												<counter>LINE</counter>
 												<value>COVEREDRATIO</value>
-												<minimum>0.9</minimum>
+												<minimum>0.7</minimum>
```
👍👍👍

kalitva commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -19,62 +20,64 @@ public interface BookingRepository extends JpaRepository<Booking, Long> {
             + "AND b.status = :status "
             + "ORDER BY b.start DESC")
-    List<Booking> findRejectedBookingsOwner(Long ownerId, BookingStatus status);
+    List<Booking> findRejectedBookingsOwner(Long ownerId, BookingStatus status, Pageable page);
```
Не помню, я говорил что все эти методы можно заменить используя `Specification`?

https://spring.io/blog/2011/04/26/advanced-spring-data-jpa-specifications-and-querydsl

kalitva commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,153 @@
+        if (itemService.findOwnerId(item.getId()).equals(bookerId)) {
+            throw new OperationAccessException("The owner cannot be a booker.");
+        }
+        if (item.getAvailable()) {
```
В таких случая обычно все невалидные кейсы идут сначала
```java
if (!item.getAvailable()) {
    throw...
}
Booking booking = Booking.builder()
...
```

kalitva commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,177 @@
+    @Override
+    @Transactional
+    public ItemDto findItemById(Long itemId, Long userId) {
+        ItemDto result;
```
Лучше объявить там же где переменная инициализируется

kalitva commented on src/test/java/ru/practicum/shareit/item/ItemControllerTest.java:
```diff
@@ -0,0 +1,253 @@
+    private ObjectMapper mapper;
+
+    @MockBean
+    private ItemService itemService;
```
В следующем задании, когда подключишь постгрес, убедись, что тесты нигде и никак не взаимодействуют с основной БД

sashajaaa committed:
Commit №15.1:
feat:
-добавлена пагинация;
-добавлена реализация запросов и ответов (Item requests);
-добавлены тесты.

refactor:
-переработан файл pom.xml;
-переработан файл schema.sql;
-переработан класс BookingController;
-переработан класс BookingMapper;
-переработан интерфейс BookingRepository;
-добавлен интерфейс BookingService;
-переработан класс BookingServiceImpl;
-переработан класс CommentMapper;
-переработан класс ItemController;
-переработан класс ItemDto;
-переработан класс ItemMapper;
-переработан класс Item;
-переработан класс ItemRepository;
-добавлен интерфейс ItemService;
-переработан класс ItemServiceImpl;
-переработан класс ItemRequestController;
-переработан класс ItemRequestDto;
-добавлен класс ItemRequestMapper;
-переработан класс ItemRequest;
-добавлен интерфейс ItemRequestRepository;
-добавлен интерфейс ItemRequestService;
-добавлен класс ItemRequestServiceImpl;
-переработан класс UserController;
-переработан класс UserMapper;
-переработан класс UserService;
-добавлен интерфейс UserService;
-переработан класс UserServiceImpl.

sashajaaa committed:
Commit №15.2:
refactor:
-переработан файл pom.xml;
-переработан класс BookingController;
-переработан интерфейс BookingService;
-переработан класс BookingServiceImpl;
-переработан интерфейс CommentRepository;
-переработан класс ItemController;
-переработан интерфейс ItemService;
-переработан класс ItemServiceImpl;
-переработан класс BookingControllerTest;
-переработан класс BookingServiceTest;
-переработан класс ItemControllerTest;
-переработан класс ItemServiceTest;
-переработан класс ItemRequestControllerTest;
-переработан класс ItemRequestBookingServiceTest;
-переработан класс UserControllerTest;
-переработан класс UserServiceTest;

sashajaaa committed:
Commit №15.2:
refactor:
-переработан файл pom.xml;
-переработан класс BookingController;
-переработан интерфейс BookingService;
-переработан класс BookingServiceImpl;
-переработан интерфейс CommentRepository;
-переработан класс ItemController;
-переработан интерфейс ItemService;
-переработан класс ItemServiceImpl;
-переработан класс BookingControllerTest;
-переработан класс BookingServiceTest;
-переработан класс ItemControllerTest;
-переработан класс ItemServiceTest;
-переработан класс ItemRequestControllerTest;
-переработан класс ItemRequestBookingServiceTest;
-переработан класс UserControllerTest;
-переработан класс UserServiceTest;

sashajaaa committed:
Commit №15.2:
refactor:
-переработан файл pom.xml;
-переработан класс BookingController;
-переработан интерфейс BookingService;
-переработан класс BookingServiceImpl;
-переработан интерфейс CommentRepository;
-переработан класс ItemController;
-переработан интерфейс ItemService;
-переработан класс ItemServiceImpl;
-переработан класс BookingControllerTest;
-переработан класс BookingServiceTest;
-переработан класс ItemControllerTest;
-переработан класс ItemServiceTest;
-переработан класс ItemRequestControllerTest;
-переработан класс ItemRequestBookingServiceTest;
-переработан класс UserControllerTest;
-переработан класс UserServiceTest;

sashajaaa committed:
Commit №15.2:
refactor:
-переработан файл pom.xml;
-переработан класс BookingController;
-переработан интерфейс BookingService;
-переработан класс BookingServiceImpl;
-переработан интерфейс CommentRepository;
-переработан класс ItemController;
-переработан интерфейс ItemService;
-переработан класс ItemServiceImpl;
-переработан класс BookingControllerTest;
-переработан класс BookingServiceTest;
-переработан класс ItemControllerTest;
-переработан класс ItemServiceTest;
-переработан класс ItemRequestControllerTest;
-переработан класс ItemRequestBookingServiceTest;
-переработан класс UserControllerTest;
-переработан класс UserServiceTest;

sashajaaa committed:
Commit №15.2:
refactor:
-переработан файл pom.xml;
-переработан класс BookingController;
-переработан интерфейс BookingService;
-переработан класс BookingServiceImpl;
-переработан интерфейс CommentRepository;
-переработан класс ItemController;
-переработан интерфейс ItemService;
-переработан класс ItemServiceImpl;
-переработан класс BookingControllerTest;
-переработан класс BookingServiceTest;
-переработан класс ItemControllerTest;
-переработан класс ItemServiceTest;
-переработан класс ItemRequestControllerTest;
-переработан класс ItemRequestBookingServiceTest;
-переработан класс UserControllerTest;
-переработан класс UserServiceTest;

kalitva commented on src/main/java/ru/practicum/shareit/booking/mapper/BookingMapper.java:
```diff
@@ -14,27 +14,25 @@
 @Component
```
Нам нужна эта аннотация?
